core/corehttp/gateway_handler.go

```diff
@@ -286,6 +293,20 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		return
 	}
 
+	// Render a 404 instead of a directory listing, if one exists
+	if resolved404Path, err404 := i.searchUpTreeFor404(r, parsedPath); err404 == nil {
+		if dr, err := i.api.Unixfs().Get(r.Context(), resolved404Path); err == nil {
+			defer dr.Close()
+
+			if f, ok := dr.(files.File); ok {
+				w.WriteHeader(http.StatusNotFound)
+				i.serveFile(w, r, "ipfs-404.html", modtime, f)
```

Comment:
In general, we should not use `serveFile` for these 404 responses:

* We should hard-code text/html as the mime type.
* We should not apply the `Range` header to 404 responses.

(and again, go doesn't like it when you call `WriteHeader` multiple times).